File: src/main/scala/Chapter2.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions.{col,explode,split, lower}

object Chapter2 {
  private val spark = SparkSession.builder()
    .appName("Chapter2")
    .master("local[*]")
    .getOrCreate()
  
  /*
    How to read a text file in Spark
    How to convert RDD to DataFrame and vice versa
    Line Count of given text file
    Word Count of given text file
  */
  
  // Through RDDs and DataFrames
  private val pathToTextFile = "./src/assets/sample.txt"

  // Through RDDs
  private val rddText = spark.sparkContext.textFile(pathToTextFile)
  // rddText.collect().foreach(println)
  // println(s"Line Count of the given file - ${rddText.collect().length}")
  
  // Thorugh DF
  private val dfText = spark.read.text(pathToTextFile)
  dfText.show()

  // Conversions

  // Convert to DF from RDD
  // println("Convert to DF from RDD")
  import spark.implicits._
  private val dfFromRDD = rddText.toDF("something_else")
  // dfFromRDD.show()

  // Convert to RDD from DF
  // println("Convert to RDD from DF")
  private val rddFromDF = dfText.rdd
  // rddFromDF.collect().foreach(println)
  
  // Word Count
  // println("Word Count of the given file")
  private val numberOfWords = rddText
  .flatMap(line => line.split(" "))
  .map(word => (word.toLowerCase, 1))
  .reduceByKey((value1,value2) => value1 + value2)
  // numberOfWords.collect().foreach(println)
  
  // Do this with dataframe

  private val dfWordCounts = dfText
                              .select(
                                explode(split(lower(col("value")), " ")).as("word")
                              )

  private val dfAggregation = dfWordCounts.groupBy("word").count()
  println("WOrd count with dataframe")
  dfAggregation.show()


  spark.stop()
}

File: src/main/python/Chapter2.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, explode, split, lower

# Initialize SparkSession
spark = SparkSession.builder.appName("Chapter2").master("local[*]").getOrCreate()

# Path to the text file
path_to_text_file = "./src/assets/sample.txt"

# Through RDDs
rdd_text = spark.sparkContext.textFile(path_to_text_file)
# Print the lines from the text file
# print("\n".join(rdd_text.collect()))

# Line count of the given file
line_count = rdd_text.count()
print(f"Line Count of the given file: {line_count}")

# Through DataFrame
df_text = spark.read.text(path_to_text_file)
df_text.show(truncate=False)

# Conversions

# Convert to DataFrame from RDD
df_from_rdd = rdd_text.toDF(["something_else"])
# df_from_rdd.show(truncate=False)

# Convert to RDD from DataFrame
rdd_from_df = df_text.rdd
# print("\n".join(map(str, rdd_from_df.collect())))

# Word Count using RDDs
rdd_word_count = rdd_text\
    .flatMap(lambda line: line.split(" "))\
        .map(lambda word: (word.lower(), 1))\
            .reduceByKey(lambda a, b: a + b)

# Print the word count
# print("\n".join(map(str, rdd_word_count.collect())))

# Word Count using DataFrame API
df_word_counts = df_text.select(
    explode(split(lower(col("value")), " ")).alias("word")
)

df_aggregation = df_word_counts.groupBy("word").count()
print("Word count with DataFrame API:")
df_aggregation.show(truncate=False)

# Stop the Spark session
spark.stop()
